internal/app/query/query_service.go

```diff
@@ -325,3 +371,146 @@ func (tc *ticketCache) update() {
+
+	// startRunRequest acts as a semaphore indicator for defining
+	// when the backfillCache is being used (queried) or is it in an idle state.
+	startRunRequest chan struct{}
+
+	// requests is the channel to handle all incoming query requests and
+	// make only one request to the storage service.
+	requests chan *cacheRequest
+
+	// This error is used for the return value when a request fails to reach
+	// the storage. In case of multiple requests are being handled, this error
+	// is returned to every single request.
+	err error
+}
+
+func newBackfillCache(b *appmain.Bindings, cfg config.View) *backfillCache {
+	bc := &backfillCache{
+		store:           statestore.New(cfg),
+		startRunRequest: make(chan struct{}, 1),
+		requests:        make(chan *cacheRequest),
+		backfills:       make(map[string]*pb.Backfill),
+	}
+
+	bc.startRunRequest <- struct{}{}
+	b.AddHealthCheckFunc(bc.store.HealthCheck)
+
+	return bc
+}
+
+func (bc *backfillCache) requestBackfills(ctx context.Context, f func(map[string]*pb.Backfill)) error {
+	cr := &cacheRequest{
+		ctx:    ctx,
+		runNow: make(chan struct{}),
+	}
+
+sendRequest:
+	for {
+		select {
+		case <-ctx.Done():
+			return errors.Wrap(ctx.Err(), "backfill cache request canceled before request sent.")
+		case <-bc.startRunRequest:
+			go bc.runRequest()
+		case bc.requests <- cr:
+			break sendRequest
+		}
+	}
+
+	select {
+	case <-ctx.Done():
+		return errors.Wrap(ctx.Err(), "backfill cache request canceled waiting for access.")
+	case <-cr.runNow:
+		defer bc.wg.Done()
+	}
+
+	if bc.err != nil {
+		return bc.err
+	}
+
+	f(bc.backfills)
+	return nil
+}
+
+func (bc *backfillCache) runRequest() {
+	defer func() {
+		bc.startRunRequest <- struct{}{}
+	}()
+
+	// Wait for first query request.
+	reqs := []*cacheRequest{<-bc.requests}
+
+	// Collect all waiting queries.
+collectAllWaiting:
+	for {
+		select {
+		case req := <-bc.requests:
+			reqs = append(reqs, req)
+		default:
+			break collectAllWaiting
+		}
+	}
+
+	bc.update()
+	stats.Record(context.Background(), cacheWaitingQueries.M(int64(len(reqs))))
+
+	// Send WaitGroup to query calls, letting them run their query on the backfill cache.
+	for _, req := range reqs {
+		select {
+		case req.runNow <- struct{}{}:
+			bc.wg.Add(1)
+		case <-req.ctx.Done():
+		}
+	}
+
+	// wait for requests to finish using backfill cache.
+	bc.wg.Wait()
+}
+
+func (bc *backfillCache) update() {
+	previousCount := len(bc.backfills)
+
+	indexedBackfills, err := bc.store.GetIndexedBackfills(context.Background())
+	if err != nil {
+		bc.err = err
+		return
+	}
+
+	deletedCount := 0
+	for id, cachedBackfill := range bc.backfills {
+		indexedBackfillGeneration, ok := indexedBackfills[id]
+		if !ok || cachedBackfill.Generation < int64(indexedBackfillGeneration) {
+			delete(bc.backfills, id)
+			deletedCount++
+		}
+	}
+
+	toFetch := []string{}
+	for id := range indexedBackfills {
+		if _, ok := bc.backfills[id]; !ok {
+			toFetch = append(toFetch, id)
```

Comment:
Why can't you call bc.store.GetBackfill right here, instead of iterating through toFetch down below? We can avoid 1 extra for loop.

Reply:
Here could be a TODO or something, in order to keep this similar to runRequest of Tickets. Later on, we can optimise it.

Reply:
Tests are flaky even without this enhancement. The main difference is `hset` vs  usual `set`.

Reply:
Here i have some doubts because, if we keep the same behavior from TicketCache, there are some metrics i need to implement and one of them uses the `toFetch` variable. So on one hand we could do what you say Alexey, or we keep the same metrics as for tickets. 
cc: @Laremere